src/openrct2-ui/windows/MapGen.cpp

```diff
@@ -374,1013 +295,1093 @@ constexpr int32_t WATERLEVEL_MIN = 0;
+    {
+        switch (page)
+        {
+            case WINDOW_MAPGEN_PAGE_BASE:
+                return BasePrepareDraw();
+            case WINDOW_MAPGEN_PAGE_RANDOM:
+                return RandomPrepareDraw();
+            case WINDOW_MAPGEN_PAGE_SIMPLEX:
+                return SimplexPrepareDraw();
+            case WINDOW_MAPGEN_PAGE_HEIGHTMAP:
+                return HeightmapPrepareDraw();
+        }
+    }
 
-static void WindowMapgenDrawTabImage(DrawPixelInfo* dpi, WindowBase* w, int32_t page, int32_t spriteIndex)
-{
-    WidgetIndex widgetIndex = WIDX_TAB_1 + page;
+    void OnDraw(DrawPixelInfo& dpi) override
+    {
+        switch (page)
+        {
+            case WINDOW_MAPGEN_PAGE_BASE:
+                return BaseDraw(dpi);
+            case WINDOW_MAPGEN_PAGE_RANDOM:
+                return RandomDraw(dpi);
+            case WINDOW_MAPGEN_PAGE_SIMPLEX:
+                return SimplexDraw(dpi);
+            case WINDOW_MAPGEN_PAGE_HEIGHTMAP:
+                return HeightmapDraw(dpi);
+        }
+    }
 
-    if (!WidgetIsDisabled(*w, widgetIndex))
+    void OnTextInput(WidgetIndex widgetIndex, std::string_view text) override
     {
-        if (w->page == page)
+        switch (page)
         {
-            int32_t frame = w->frame_no / TabAnimationDivisor[w->page];
-            spriteIndex += (frame % TabAnimationFrames[w->page]);
+            case WINDOW_MAPGEN_PAGE_BASE:
+                return TextInput(widgetIndex, text);
+            case WINDOW_MAPGEN_PAGE_SIMPLEX:
+                return TextInput(widgetIndex, text);
         }
+    }
+
+    void AfterLoadingHeightMap(int32_t result, const utf8* path)
+    {
+        if (result == MODAL_RESULT_OK)
+        {
+            if (!MapGenLoadHeightmap(path))
+            {
+                // TODO: Display error popup
+                return;
+            }
+
+            // The window needs to be open while using the map
+            _heightmapLoaded = true;
+            SetPage(WINDOW_MAPGEN_PAGE_HEIGHTMAP);
 
-        GfxDrawSprite(
-            dpi, ImageId(spriteIndex),
-            w->windowPos + ScreenCoordsXY{ w->widgets[widgetIndex].left, w->widgets[widgetIndex].top });
+            HeightmapGenerateMap();
+        }
     }
-}
+};
 
-static void WindowMapgenDrawTabImages(DrawPixelInfo* dpi, WindowBase* w)
+WindowBase* WindowMapgenOpen()
 {
-    WindowMapgenDrawTabImage(dpi, w, WINDOW_MAPGEN_PAGE_BASE, SPR_G2_TAB_LAND);
-    WindowMapgenDrawTabImage(dpi, w, WINDOW_MAPGEN_PAGE_RANDOM, SPR_G2_TAB_TREE);
-    WindowMapgenDrawTabImage(dpi, w, WINDOW_MAPGEN_PAGE_SIMPLEX, SPR_G2_TAB_PENCIL);
-    WindowMapgenDrawTabImage(dpi, w, WINDOW_MAPGEN_PAGE_HEIGHTMAP, SPR_TAB_GRAPH_0);
+    return WindowFocusOrCreate<MapGenWindow>(WindowClass::Mapgen, WW, WH, WF_10 | WF_AUTO_POSITION | WF_CENTRE_SCREEN);
 }
 
-#pragma endregion
+static void HeightmapLoadsaveCallback(int32_t result, const utf8* path)
+{
+    auto* w = static_cast<MapGenWindow*>(WindowMapgenOpen());
```

Comment:
This is a tiny change from previous in that it will call focus on the window. It probably makes sense though.

Reply:
I tested it with both a correct and incorrect heightmap, it appeared to behave correctly in both cases.